File: 330/2B.rb
#!/usr/bin/ruby

n, k = gets.strip().split(" ")
n = n.to_i
k = k.to_i
a = gets.strip().split(" ")
a = a.map(&:to_i)
b = gets.strip().split(" ")
b = b.map(&:to_i)
m = n / k;
ans = 1;
for i in 0..m - 1
    if b[i] == 0  
        tmp = 10 ** (k - 1)
        mul = (tmp * 10 - 1) / a[i] - tmp / a[i] + 1
        if tmp % a[i] > 0
            mul -= 1
        end
    else
        tmp = 10 ** (k - 1) 
        mul = (tmp * 10 - 1) / a[i] - tmp * (b[i] + 1) / a[i] + 1 + (tmp * b[i] - 1) / a[i] + 1
        if tmp * (b[i] + 1) % a[i] > 0
            mul -= 1
        end
    end
    ans = ans * mul
    ans %= 10 ** 9 + 7
end
print ans, "\n"

File: 330/2B.php
<?php
fscanf(STDIN, "%d %d", $n, $k);
$temp = trim(fgets(STDIN));
$a = explode(" ", $temp);
$temp = trim(fgets(STDIN));
$b = explode(" ", $temp);
$m = $n / $k;
$ans = 1;
for($i = 0;$i < $m;$i ++) {
    if($b[$i] == 0) {
        $tmp = pow(10, $k - 1);
        $mul = intval(($tmp * 10 - 1) / $a[$i]) - intval($tmp / $a[$i]) + 1;
        if ($tmp % $a[$i]) $mul --;
    } else {
        $tmp = pow(10, $k - 1);
        $mul = intval(($tmp * 10 - 1) / $a[$i]) - intval($tmp * ($b[$i] + 1) / $a[$i]) + 1 + intval(($tmp * $b[$i] - 1) / $a[$i]) + 1;
        if ($tmp * ($b[$i] + 1) % $a[$i]) $mul --;
    }
    $ans = $ans * $mul;
    $ans %= (pow(10, 9) + 7);
}
echo intval($ans), "\n";
?>